File: skills/langsmith-evaluator/scripts/upload_evaluators.py
#!/usr/bin/env python3
"""Upload evaluators to LangSmith."""

import inspect
import os
from collections.abc import Callable
from dataclasses import dataclass

import click
import requests
from dotenv import load_dotenv
from langsmith import Client
from rich.console import Console
from rich.table import Table

load_dotenv()

console = Console()

# Configuration
LANGSMITH_API_KEY = os.getenv("LANGSMITH_API_KEY")
LANGSMITH_API_URL = os.getenv("LANGSMITH_API_URL", "https://api.smith.langchain.com")
LANGSMITH_WORKSPACE_ID = os.getenv("LANGSMITH_WORKSPACE_ID")

if not LANGSMITH_API_KEY:
    raise ValueError("LANGSMITH_API_KEY environment variable is required")


@dataclass
class CodeEvaluator:
    """Code-based evaluator configuration."""

    code: str
    language: str = "python"


@dataclass
class EvaluatorPayload:
    """Evaluator upload payload."""

    display_name: str
    evaluators: list[CodeEvaluator]
    sampling_rate: float
    target_dataset_ids: list[str] | None = None
    target_project_ids: list[str] | None = None


def get_headers() -> dict:
    """Get API headers with authentication."""
    headers = {"x-api-key": LANGSMITH_API_KEY, "Content-Type": "application/json"}
    if LANGSMITH_WORKSPACE_ID:
        headers["x-tenant-id"] = LANGSMITH_WORKSPACE_ID
    return headers


def evaluator_exists(name: str) -> bool:
    """Check if evaluator already exists."""
    url = f"{LANGSMITH_API_URL}/runs/rules"
    response = requests.get(url, headers=get_headers())
    response.raise_for_status()

    rules = response.json()
    return any(rule.get("display_name") == name for rule in rules)


def resolve_dataset_id(dataset_name: str) -> str | None:
    """Get dataset ID from name."""
    try:
        client = Client()
        dataset = client.read_dataset(dataset_name=dataset_name)
        return str(dataset.id)
    except Exception as e:
        console.print(f"[yellow]Warning: Could not find dataset '{dataset_name}': {e}[/yellow]")
        return None


def resolve_project_id(project_name: str) -> str | None:
    """Get project ID from name."""
    try:
        client = Client()
        # List projects and find matching name
        for project in client.list_projects():
            if project.name == project_name:
                return str(project.id)
        console.print(f"[yellow]Warning: Could not find project '{project_name}'[/yellow]")
        return None
    except Exception as e:
        console.print(f"[yellow]Warning: Error finding project '{project_name}': {e}[/yellow]")
        return None


def create_code_payload(
    name: str,
    func: Callable,
    sample_rate: float = 1.0,
    target_dataset: str | None = None,
    target_project: str | None = None,
    replace: bool = False,
    skip_confirm: bool = False,
) -> EvaluatorPayload | None:
    """Create payload for code-based evaluator.

    Args:
        name: Display name for evaluator
        func: Python function with any name - will be renamed to perform_eval for LangSmith
        sample_rate: Sampling rate (0.0-1.0)
        target_dataset: Optional dataset name to attach to
        target_project: Optional project name to attach to
        replace: If True, delete existing evaluator first
    """
    # Check if exists
    if evaluator_exists(name):
        if not replace:
            console.print(
                f"[yellow]Evaluator '{name}' already exists. Use --replace to overwrite.[/yellow]"
            )
            return None
        else:
            if not skip_confirm:
                console.print(f"[yellow]⚠️  Evaluator '{name}' already exists.[/yellow]")
                response_text = input("Replace existing evaluator? (y/n): ").lower().strip()
                if response_text != "y":
                    console.print("[yellow]Upload cancelled[/yellow]")
                    return None
            delete_evaluator(name, confirm=False)  # Already confirmed above or skipped

    # Get function source code
    source = inspect.getsource(func)

    # LangSmith API requires function to be named 'perform_eval'
    # Replace the function name in the source code
    func_name = func.__name__
    import re

    # Replace "def function_name(" with "def perform_eval("
    source = re.sub(rf"\bdef\s+{re.escape(func_name)}\s*\(", "def perform_eval(", source)

    # Create evaluator
    code_evaluator = CodeEvaluator(code=source, language="python")

    # Resolve targets
    dataset_ids = []
    project_ids = []

    if target_dataset:
        dataset_id = resolve_dataset_id(target_dataset)
        if dataset_id:
            dataset_ids.append(dataset_id)

    if target_project:
        project_id = resolve_project_id(target_project)
        if project_id:
            project_ids.append(project_id)

    return EvaluatorPayload(
        display_name=name,
        evaluators=[code_evaluator],
        sampling_rate=sample_rate,
        target_dataset_ids=dataset_ids if dataset_ids else None,
        target_project_ids=project_ids if project_ids else None,
    )


def delete_evaluator(name: str, confirm: bool = True) -> bool:
    """Delete an evaluator by name.

    Args:
        name: Evaluator name
        confirm: If True, prompt for confirmation before deleting
    """
    # Get all rules
    url = f"{LANGSMITH_API_URL}/runs/rules"
    response = requests.get(url, headers=get_headers())
    response.raise_for_status()

    rules = response.json()
    rule = next((r for r in rules if r.get("display_name") == name), None)

    if not rule:
        console.print(f"[yellow]Evaluator '{name}' not found[/yellow]")
        return False

    # Confirm deletion
    if confirm:
        console.print(f"[yellow]⚠️  About to delete evaluator: '{name}'[/yellow]")
        response_text = input("Are you sure? (y/n): ").lower().strip()
        if response_text != "y":
            console.print("[yellow]Deletion cancelled[/yellow]")
            return False

    # Delete the rule
    rule_id = rule.get("id")
    delete_url = f"{LANGSMITH_API_URL}/runs/rules/{rule_id}"
    response = requests.delete(delete_url, headers=get_headers())
    response.raise_for_status()

    console.print(f"[green]✓ Deleted evaluator '{name}'[/green]")
    return True


def create_evaluator(payload: EvaluatorPayload) -> bool:
    """Upload evaluator to LangSmith."""
    url = f"{LANGSMITH_API_URL}/runs/rules"

    # Convert payload to dict - use code_evaluators for code-based evaluators
    data = {
        "display_name": payload.display_name,
        "sampling_rate": payload.sampling_rate,
        "code_evaluators": [{"code": e.code, "language": e.language} for e in payload.evaluators],
    }

    if payload.target_dataset_ids:
        data["dataset_id"] = (
            payload.target_dataset_ids[0] if len(payload.target_dataset_ids) == 1 else None
        )
    if payload.target_project_ids:
        data["session_id"] = (
            payload.target_project_ids[0] if len(payload.target_project_ids) == 1 else None
        )

    # Upload
    response = requests.post(url, json=data, headers=get_headers())

    if response.status_code == 200:
        console.print(f"[green]✓ Uploaded evaluator '{payload.display_name}'[/green]")
        return True
    else:
        console.print(f"[red]✗ Failed to upload '{payload.display_name}': {response.text}[/red]")
        return False


@click.group()
def cli():
    """Upload and manage LangSmith evaluators."""
    pass


@cli.command()
def list():
    """List all evaluators."""
    url = f"{LANGSMITH_API_URL}/runs/rules"
    response = requests.get(url, headers=get_headers())
    response.raise_for_status()

    rules = response.json()

    if not rules:
        console.print("[yellow]No evaluators found[/yellow]")
        return

    table = Table(title="LangSmith Evaluators")
    table.add_column("Name", style="cyan")
    table.add_column("Sampling Rate", style="green")
    table.add_column("Targets", style="yellow")

    for rule in rules:
        name = rule.get("display_name", "")
        rate = rule.get("sampling_rate", 1.0)

        targets = []
        if rule.get("target_dataset_ids"):
            targets.append(f"{len(rule['target_dataset_ids'])} datasets")
        if rule.get("target_project_ids"):
            targets.append(f"{len(rule['target_project_ids'])} projects")

        table.add_row(name, f"{rate:.1%}", ", ".join(targets) if targets else "All runs")

    console.print(table)


@cli.command()
@click.argument("name")
@click.option("--yes", is_flag=True, help="Skip confirmation prompt")
def delete(name: str, yes: bool):
    """Delete an evaluator by name."""
    delete_evaluator(name, confirm=not yes)


@cli.command()
@click.argument("evaluator_file")
@click.option("--name", required=True, help="Display name for evaluator")
@click.option("--function", required=True, help="Function name to extract from file")
@click.option("--dataset", help="Target dataset name")
@click.option("--project", help="Target project name")
@click.option("--sample-rate", default=1.0, type=float, help="Sampling rate (0.0-1.0)")
@click.option("--replace", is_flag=True, help="Replace if exists")
@click.option("--yes", is_flag=True, help="Skip confirmation prompts")
def upload(
    evaluator_file: str,
    name: str,
    function: str,
    dataset: str | None,
    project: str | None,
    sample_rate: float,
    replace: bool,
    yes: bool,
):
    """Upload an evaluator from a Python file.

    Example:
        python upload_evaluators.py upload my_evaluators.py \\
            --name "Trajectory Match" \\
            --function trajectory_match_evaluator \\
            --dataset "Skills: Trajectory" \\
            --replace
    """
    # Load the Python file
    import importlib.util

    spec = importlib.util.spec_from_file_location("evaluators", evaluator_file)
    if not spec or not spec.loader:
        console.print(f"[red]Failed to load {evaluator_file}[/red]")
        return

    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    # Get the function
    if not hasattr(module, function):
        console.print(f"[red]Function '{function}' not found in {evaluator_file}[/red]")
        return

    func = getattr(module, function)

    # Create payload
    payload = create_code_payload(
        name=name,
        func=func,
        sample_rate=sample_rate,
        target_dataset=dataset,
        target_project=project,
        replace=replace,
        skip_confirm=yes,
    )

    if payload:
        create_evaluator(payload)


if __name__ == "__main__":
    cli()

File: skills/langsmith-evaluator/scripts/upload_evaluators.ts
#!/usr/bin/env npx tsx
/**
 * Upload JavaScript evaluators to LangSmith.
 *
 * This TypeScript CLI uploads JavaScript code evaluators.
 * For Python evaluators, use upload_evaluators.py instead.
 */

import { Client } from "langsmith";
import { Command } from "commander";
import chalk from "chalk";
import Table from "cli-table3";
import * as dotenv from "dotenv";
import * as fs from "fs";
import * as readline from "readline";

dotenv.config();

// ============================================================================
// Configuration
// ============================================================================

export const LANGSMITH_API_KEY = process.env.LANGSMITH_API_KEY;
export const LANGSMITH_API_URL =
  process.env.LANGSMITH_API_URL || "https://api.smith.langchain.com";
export const LANGSMITH_WORKSPACE_ID = process.env.LANGSMITH_WORKSPACE_ID;

// Only validate API key when running as CLI (not when imported for testing)
const isMainModule =
  process.argv[1] &&
  (process.argv[1].endsWith("upload_evaluators.ts") ||
    process.argv[1].endsWith("upload_evaluators.js"));

if (isMainModule && !LANGSMITH_API_KEY) {
  console.error(
    chalk.red("Error: LANGSMITH_API_KEY environment variable is required"),
  );
  process.exit(1);
}

// ============================================================================
// Types
// ============================================================================

export interface CodeEvaluator {
  code: string;
  language: string;
}

export interface EvaluatorPayload {
  display_name: string;
  evaluators: CodeEvaluator[];
  sampling_rate: number;
  target_dataset_ids?: string[];
  target_project_ids?: string[];
}

export interface Rule {
  id: string;
  display_name: string;
  sampling_rate: number;
  target_dataset_ids?: string[];
  target_project_ids?: string[];
}

// ============================================================================
// API Helpers
// ============================================================================

export function getHeaders(): Record<string, string> {
  const headers: Record<string, string> = {
    "x-api-key": LANGSMITH_API_KEY!,
    "Content-Type": "application/json",
  };
  if (LANGSMITH_WORKSPACE_ID) {
    headers["x-tenant-id"] = LANGSMITH_WORKSPACE_ID;
  }
  return headers;
}

export async function getRules(): Promise<Rule[]> {
  const url = `${LANGSMITH_API_URL}/runs/rules`;
  const response = await fetch(url, { headers: getHeaders() });
  if (!response.ok) {
    throw new Error(`Failed to get rules: ${response.statusText}`);
  }
  return response.json() as Promise<Rule[]>;
}

export async function evaluatorExists(name: string): Promise<boolean> {
  const rules = await getRules();
  return rules.some((rule) => rule.display_name === name);
}

async function resolveDatasetId(datasetName: string): Promise<string | null> {
  try {
    const client = new Client({ apiKey: LANGSMITH_API_KEY });
    const dataset = await client.readDataset({ datasetName });
    return dataset.id;
  } catch (e) {
    console.log(
      chalk.yellow(`Warning: Could not find dataset '${datasetName}': ${e}`),
    );
    return null;
  }
}

async function resolveProjectId(projectName: string): Promise<string | null> {
  try {
    const client = new Client({ apiKey: LANGSMITH_API_KEY });
    for await (const project of client.listProjects()) {
      if (project.name === projectName) {
        return project.id;
      }
    }
    console.log(
      chalk.yellow(`Warning: Could not find project '${projectName}'`),
    );
    return null;
  } catch (e) {
    console.log(
      chalk.yellow(`Warning: Error finding project '${projectName}': ${e}`),
    );
    return null;
  }
}

// ============================================================================
// Evaluator Operations
// ============================================================================

async function promptConfirm(message: string): Promise<boolean> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    rl.question(`${message} (y/n): `, (answer) => {
      rl.close();
      resolve(answer.toLowerCase().trim() === "y");
    });
  });
}

export async function deleteEvaluator(
  name: string,
  confirm = true,
): Promise<boolean> {
  const rules = await getRules();
  const rule = rules.find((r) => r.display_name === name);

  if (!rule) {
    console.log(chalk.yellow(`Evaluator '${name}' not found`));
    return false;
  }

  if (confirm) {
    console.log(chalk.yellow(`About to delete evaluator: '${name}'`));
    const confirmed = await promptConfirm("Are you sure?");
    if (!confirmed) {
      console.log(chalk.yellow("Deletion cancelled"));
      return false;
    }
  }

  const deleteUrl = `${LANGSMITH_API_URL}/runs/rules/${rule.id}`;
  const response = await fetch(deleteUrl, {
    method: "DELETE",
    headers: getHeaders(),
  });

  if (!response.ok) {
    throw new Error(`Failed to delete: ${response.statusText}`);
  }

  console.log(chalk.green(`✓ Deleted evaluator '${name}'`));
  return true;
}

function extractJavaScriptFunction(
  fileContent: string,
  functionName: string,
): string | null {
  // Match JavaScript/TypeScript function definitions
  // Handles: function name(), async function name(), const name = () =>, const name = async () =>
  const patterns = [
    // async function functionName(...)  { ... }
    new RegExp(
      `(async\\s+function\\s+${functionName}\\s*\\([^)]*\\)\\s*\\{[\\s\\S]*?\\n\\})`,
      "m",
    ),
    // function functionName(...) { ... }
    new RegExp(
      `(function\\s+${functionName}\\s*\\([^)]*\\)\\s*\\{[\\s\\S]*?\\n\\})`,
      "m",
    ),
    // const functionName = async (...) => { ... }
    new RegExp(
      `(const\\s+${functionName}\\s*=\\s*async\\s*\\([^)]*\\)\\s*=>\\s*\\{[\\s\\S]*?\\n\\})`,
      "m",
    ),
    // const functionName = (...) => { ... }
    new RegExp(
      `(const\\s+${functionName}\\s*=\\s*\\([^)]*\\)\\s*=>\\s*\\{[\\s\\S]*?\\n\\})`,
      "m",
    ),
  ];

  for (const pattern of patterns) {
    const match = fileContent.match(pattern);
    if (match) {
      let source = match[1].trim();
      // Rename function to performEval as required by LangSmith for JS
      source = source.replace(
        new RegExp(`(async\\s+)?function\\s+${functionName}\\s*\\(`),
        "$1function performEval(",
      );
      source = source.replace(
        new RegExp(`const\\s+${functionName}\\s*=`),
        "function performEval",
      );
      // Handle arrow function conversion to regular function
      source = source.replace(
        /function performEval\s*async\s*\(([^)]*)\)\s*=>\s*\{/,
        "async function performEval($1) {",
      );
      source = source.replace(
        /function performEval\s*\(([^)]*)\)\s*=>\s*\{/,
        "function performEval($1) {",
      );
      return source;
    }
  }

  return null;
}

async function createCodePayload(options: {
  name: string;
  source: string;
  sampleRate: number;
  targetDataset?: string;
  targetProject?: string;
  replace: boolean;
  skipConfirm: boolean;
}): Promise<EvaluatorPayload | null> {
  // Check if exists
  if (await evaluatorExists(options.name)) {
    if (!options.replace) {
      console.log(
        chalk.yellow(
          `Evaluator '${options.name}' already exists. Use --replace to overwrite.`,
        ),
      );
      return null;
    } else {
      if (!options.skipConfirm) {
        console.log(
          chalk.yellow(`Evaluator '${options.name}' already exists.`),
        );
        const confirmed = await promptConfirm("Replace existing evaluator?");
        if (!confirmed) {
          console.log(chalk.yellow("Upload cancelled"));
          return null;
        }
      }
      await deleteEvaluator(options.name, false);
    }
  }

  // Resolve targets
  const datasetIds: string[] = [];
  const projectIds: string[] = [];

  if (options.targetDataset) {
    const datasetId = await resolveDatasetId(options.targetDataset);
    if (datasetId) {
      datasetIds.push(datasetId);
    }
  }

  if (options.targetProject) {
    const projectId = await resolveProjectId(options.targetProject);
    if (projectId) {
      projectIds.push(projectId);
    }
  }

  return {
    display_name: options.name,
    evaluators: [{ code: options.source, language: "javascript" }],
    sampling_rate: options.sampleRate,
    target_dataset_ids: datasetIds.length > 0 ? datasetIds : undefined,
    target_project_ids: projectIds.length > 0 ? projectIds : undefined,
  };
}

export async function createEvaluator(
  payload: EvaluatorPayload,
): Promise<boolean> {
  const url = `${LANGSMITH_API_URL}/runs/rules`;

  const data: Record<string, unknown> = {
    display_name: payload.display_name,
    sampling_rate: payload.sampling_rate,
    code_evaluators: payload.evaluators.map((e) => ({
      code: e.code,
      language: e.language,
    })),
  };

  if (payload.target_dataset_ids && payload.target_dataset_ids.length === 1) {
    data.dataset_id = payload.target_dataset_ids[0];
  }
  if (payload.target_project_ids && payload.target_project_ids.length === 1) {
    data.session_id = payload.target_project_ids[0];
  }

  const response = await fetch(url, {
    method: "POST",
    headers: getHeaders(),
    body: JSON.stringify(data),
  });

  if (response.ok) {
    console.log(chalk.green(`✓ Uploaded evaluator '${payload.display_name}'`));
    return true;
  } else {
    const text = await response.text();
    console.log(
      chalk.red(`✗ Failed to upload '${payload.display_name}': ${text}`),
    );
    return false;
  }
}

// ============================================================================
// CLI
// ============================================================================

const program = new Command();

program
  .name("upload_evaluators")
  .description("Upload and manage JavaScript evaluators in LangSmith")
  .version("1.0.0");

// list command
program
  .command("list")
  .description("List all evaluators")
  .action(async () => {
    const rules = await getRules();

    if (rules.length === 0) {
      console.log(chalk.yellow("No evaluators found"));
      return;
    }

    const table = new Table({
      head: [
        chalk.bold("Name"),
        chalk.bold("Sampling Rate"),
        chalk.bold("Targets"),
      ],
      style: { head: [], border: [] },
    });

    for (const rule of rules) {
      const name = rule.display_name || "";
      const rate = rule.sampling_rate || 1.0;

      const targets: string[] = [];
      if (rule.target_dataset_ids && rule.target_dataset_ids.length > 0) {
        targets.push(`${rule.target_dataset_ids.length} datasets`);
      }
      if (rule.target_project_ids && rule.target_project_ids.length > 0) {
        targets.push(`${rule.target_project_ids.length} projects`);
      }

      table.push([
        chalk.cyan(name),
        chalk.green(`${(rate * 100).toFixed(0)}%`),
        chalk.yellow(targets.length > 0 ? targets.join(", ") : "All runs"),
      ]);
    }

    console.log(chalk.bold("LangSmith Evaluators"));
    console.log(table.toString());
  });

// delete command
program
  .command("delete <name>")
  .description("Delete an evaluator by name")
  .option("--yes", "Skip confirmation prompt")
  .action(async (name, opts) => {
    await deleteEvaluator(name, !opts.yes);
  });

// upload command
program
  .command("upload <evaluatorFile>")
  .description("Upload a JavaScript evaluator from a .js or .ts file")
  .requiredOption("--name <name>", "Display name for evaluator")
  .requiredOption("--function <name>", "Function name to extract from file")
  .option("--dataset <name>", "Target dataset name")
  .option("--project <name>", "Target project name")
  .option("--sample-rate <rate>", "Sampling rate (0.0-1.0)", "1.0")
  .option("--replace", "Replace if exists")
  .option("--yes", "Skip confirmation prompts")
  .action(async (evaluatorFile, opts) => {
    // Read the file
    if (!fs.existsSync(evaluatorFile)) {
      console.log(chalk.red(`File not found: ${evaluatorFile}`));
      return;
    }

    const fileContent = fs.readFileSync(evaluatorFile, "utf-8");

    // Extract the function source
    const source = extractJavaScriptFunction(fileContent, opts.function);
    if (!source) {
      console.log(
        chalk.red(`Function '${opts.function}' not found in ${evaluatorFile}`),
      );
      return;
    }

    // Create payload
    const payload = await createCodePayload({
      name: opts.name,
      source,
      sampleRate: parseFloat(opts.sampleRate),
      targetDataset: opts.dataset,
      targetProject: opts.project,
      replace: opts.replace || false,
      skipConfirm: opts.yes || false,
    });

    if (payload) {
      await createEvaluator(payload);
    }
  });

if (isMainModule) {
  program.parse();
}
